Reconcile iptables rules in case of runtime device updates

pippolo84 opened the pull request:
iptables rules manager currently exposes some methods to reinstall rules after a datapath reconfiguration (`loader.Reinitialize`).

This PR changes the manager so that it dynamically reconcile the iptables rules every time:

- there is an update regarding a device
- there is an update for the local node
- a new proxy redirection is installed
- an ipset is changed (addition/removal of an ip in the set)
- a pod with "policy.cilium.io/no-track-port" annotation is added or removed

The exposed methods are changed so that each request is processed asynchronously by the manager internal reconciler.

pippolo84 committed:
iptables: Manage rules through runtime reconciliation

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Remove usage of check pkg in tests

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Add test for reconciliation loop

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,72 @@
+		installRules: installRules,
+	}
+
+	localNode, err := localNodeStore.Get(ctx)
```
This isn't needed, `localNodeStore.Observable` emits the current state on subscribe, so you can just do `state.localNode = <-localNodeEvents` before the loop.

joamaki commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,72 @@
+		}
+	}
+
+	health.Stopped("Context done")
```
jobs will call `Stopped` for you

joamaki commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,72 @@
+		txn := db.ReadTxn()
+		var devicesWatch <-chan struct{}
+		state.devices, devicesWatch = tables.SelectedDevices(devices, txn)
+
```
Might be worth adding rate-limiting here to do this e.g. max once per second (e.g. `rate.NewLimiter(time.Second, 1)`)

joamaki commented on pkg/datapath/iptables/reconciler_test.go:
```diff
@@ -0,0 +1,241 @@
+		}
+		current.devices = make([]*tables.Device, 0, len(state.devices))
+		for _, dev := range state.devices {
+			current.devices = append(current.devices, dev.DeepCopy())
```
No need to DeepCopy, they're immutable

joamaki commented on pkg/datapath/iptables/reconciler_test.go:
```diff
@@ -0,0 +1,241 @@
+		for _, dev := range state.devices {
+			current.devices = append(current.devices, dev.DeepCopy())
+		}
+		current.localNode = node.LocalNode{Node: *state.localNode.DeepCopy()}
```
No need to DeepCopy

joamaki commented on pkg/datapath/iptables/reconciler_test.go:
```diff
@@ -0,0 +1,241 @@
+		),
+	)
+
+	states := make(chan *desiredState)
```
Using a channel here seems dangerous since it requires that the reconciler does exactly one call to the update, but that may not be the case if it's fed multiple pieces of data (like we are via local node store and the devices table). Instead you could consider doing an atomic pointer swap and use `Eventually` to check that we reach the right state. That way we don't accidentally block the reconcilation loop.

joamaki commented on pkg/datapath/iptables/reconciler_test.go:
```diff
@@ -0,0 +1,241 @@
+	close(states)
+
+	// drain the states channel
+	for range states {
```
Drain after `close` is a no-op. Also is it safe to `close` above? Do we really know nothing will send to it?

pippolo84 replied:
This is a leftover from a previous version, I'll remove it. :+1:

pippolo84 committed:
iptables: Fix `New port number` case in TestAddProxyRules{v4,v6}

The third part of the `TestAddProxyRulesv4` has a comment stating:

// New port number, adds new ones, deletes stale rules. Does not touch OLD_ chains

But it actually uses the same port number as the first part of the test,
starting from the same installed iptables rules. As a result, it is not
actually testing the rules updating after a call to addProxyRules with a
new proxy port.

The test case has been fixed to start from a redirection already
installed for port "37379" and then calling addProxyRules to change the
port to "37380". Doing that the test case exercises the ability of the
code to update the rules when a new port is used for the proxy.

The same fix has been applied for the similar case in
TestAddProxyRulesv6.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Extract runnable interface from iptablesInterface

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Add tests for ipset list

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
Debug e2e test

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,197 @@
+	ctx context.Context,
+	health cell.HealthReporter,
+	installIptRules bool,
+	dependents *dependents,
```
let's call this `reconcilerParams` instead for consistent naming (e.g. params is something a cell gets from outside).

joamaki commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,197 @@
+			state.localNodeInfo = localNodeInfo
+		case proxyInfo, ok := <-dependents.proxies:
+			if !ok {
+				return nil
```
we're not draining the other channels when we return here which may cause them to block on send. For example `localNodeEvents` is unbuffered so bailing out here without draining and cancelling the context will block `LocalNodeStore` completely. Please use `context.WithCancel` and add defer'd cancelling and draining of all (external) channels.

joamaki commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,197 @@
+			delete(state.noTrackPods, noTrackPod)
+		}
+
+		log.WithField("state", state).Info("updating rules to reconcile state")
```
I'm not sure it's a good idea to dump the whole state in logs (potentially once per second). How about we log the whole state if debug level is on and otherwise a condensed summary (e.g. number of entries and so on)?

pippolo84 replied:
Yeah, this is just an additional log that I added in a separate commit while trying to debug the ci-e2e failure. I'm gonna remove it :+1:

joamaki commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,197 @@
+		log.WithField("state", state).Info("updating rules to reconcile state")
+
+		if err := updateRules(ctx, state, false); err != nil {
+			health.Degraded("iptables rules update failed", err)
```
If this fails we want to retry with backoff. See `pkg/service/reconciler.go`.

joamaki replied:
The retrying from `updateRules` should move here as well. We want to be able to receive updates to data while we retry.

joamaki commented on pkg/datapath/iptables/cell.go:
```diff
@@ -26,6 +26,7 @@ var Cell = cell.Module(
 			IptablesMasqueradingIPv4Enabled: cfg.IptablesMasqueradingIPv4Enabled(),
 			IptablesMasqueradingIPv6Enabled: cfg.IptablesMasqueradingIPv6Enabled(),
 			IPv4NativeRoutingCIDR:           cfg.GetIPv4NativeRoutingCIDR(),
```
I'm not sure these are available yet at the time this is called (take a look where SetIPv4NativeRoutingCIDR is called from). We might potentially need to do some fixes to this...

pippolo84 replied:
Fixed in https://github.com/cilium/cilium/pull/30918

joamaki committed:
iptables: Use a ticker to decide when to reconcile

The rate limiting effectively limited how quickly the ipset etc. changes were
processed. Use a ticker instead to set the rate at which the changed desired
state is reconciled. Additionally move the retrying to be done via the ticker
as well so new changes to desired state are taken into account when retrying.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
